### PR TITLE
update code to use original image and video pixel coordinates for all predictions

### PR DESCRIPTION
**Pull request by Selmaan**
this applies to supervised & unsupervised training, and inference. Image size is also no longer part of a configuration. Image sizes are always computed on the fly, and passed as an entry in batch_dicts (bbox). This is used to transform predictions from resized to original coords. This should make the code long-term compatible with upcoming changes for a 2-stage dynamic-crop algorithm. At present, it only affects the code by rescaling all losses computed per-pixel, e.g. a temporal loss threshold of 20 pixels no applies to 20 pixels in the original, not resized, coords.

### CODE REVIEW

File: lightning_pose/data/datasets.py [hunk 1]
```diff
@@ -260,8 +261,6 @@ def __init__(
         # Compute heatmaps as preprocessing step
         self.num_targets = torch.numel(self.keypoints[0])
         self.num_keypoints = self.num_targets // 2
-        self.label_heatmaps = None  # populated by `self.compute_heatmaps()`
-        self.compute_heatmaps()
```
Comment (themattinthehatt) [inline lightning_pose/data/datasets.py:264]: nice

File: lightning_pose/data/datasets.py [hunk 2]
```diff
@@ -328,10 +327,156 @@ def __getitem__(self, idx: int) -> HeatmapLabeledExampleDict:
 
         """
         example_dict: BaseLabeledExampleDict = super().__getitem__(idx)
-        if len(self.imgaug_transform) == 1 and isinstance(self.imgaug_transform[0], iaa.Resize):
-            # we have a deterministic resizing augmentation; use precomputed heatmaps
-            example_dict["heatmaps"] = self.label_heatmaps[idx]
```
Comment (themattinthehatt) [inline lightning_pose/data/datasets.py:333]: thanks for ditching this, should have happened long ago

File: lightning_pose/data/datasets.py [hunk 3]
```diff
@@ -328,10 +327,156 @@ def __getitem__(self, idx: int) -> HeatmapLabeledExampleDict:
...
+        example_dict["heatmaps"] = self.compute_heatmap(example_dict)
+        return example_dict
+
+
+class DetectorDataset(BaseTrackingDataset):
```
Comment (themattinthehatt) [inline lightning_pose/data/datasets.py:334]: cleaner if you inherit from `HeatmapDataset` - then you don't need to to redefine `compute_heatmap`, `output_shape`, etc., as well as some of the attribute setting in `__init__` (`output_sigma`, `uniform_heatmaps`, etc.).

you can still overwrite `__getitem__` to do whatever you want

File: lightning_pose/models/heatmap_tracker.py [hunk 4]
```diff
@@ -269,16 +269,40 @@ def forward(
...
+            predicted_keypoints[:, :, 1] /= batch_dict["images"].shape[-2]  # -2 dim is height "y"
+        else:  # we have unlabeled dict, 'frames' instead of 'images'
+            predicted_keypoints[:, :, 0] /= batch_dict["frames"].shape[-1]  # -1 dim is width "x"
+            predicted_keypoints[:, :, 1] /= batch_dict["frames"].shape[-2]  # -2 dim is height "y"
+        # multiply and add by bbox dims (x,y,h,w)
+        for i in range(predicted_keypoints.shape[0]):
+            predicted_keypoints[i, :, 0] *= batch_dict["bbox"][i, 3]  # scale x by box width
+            predicted_keypoints[i, :, 0] += batch_dict["bbox"][i, 0]  # add bbox x offset
+            predicted_keypoints[i, :, 1] *= batch_dict["bbox"][i, 2]  # scale y by box height
+            predicted_keypoints[i, :, 1] += batch_dict["bbox"][i, 1]  # add bbox y offset
```
Comment (themattinthehatt) [inline lightning_pose/models/heatmap_tracker.py:289]: can this be vectorized? should check if something like `predicted_keypoints[:, :, 0] *= batch_dict["bbox"][:, 3]` works (might need `batch_dict["bbox"][:, 3].unsqueeze(1)` or something)

File: scripts/configs/config_birdCOM.yaml [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+  # max allowed number of instances, i.e. 1 for single-animal, >1 for multiple animal
+  num_max_instances: 1
+  # detector network params
+
+training:
+  # select from one of several predefined image/video augmentation pipelines
+  # default- resizing only
+  # dlc- imgaug pipeline implemented in DLC 2.0 package
+  # dlc-top-down- dlc augmentations plus vertical and horizontal flips
+  imgaug: dlc-top-down
```
Comment (themattinthehatt) [inline scripts/configs/config_birdCOM.yaml:32]: fyi for some reason i found `dlc-top-down` to be worse than `dlc` for a top-down two-mouse dataset - later on i'd be interested in how these compare with your dataset

**Review by themattinthehatt — CHANGES_REQUESTED** [review]
@Selmaan PR looks great! I added some comments, take a look and let me know if you have any questions. besides the comments it would be great if you could add a unit test for `convert_bbox_coords` - then I think we can wrap this up

File: lightning_pose/data/augmentations.py [hunk 6]
```diff
@@ -97,11 +97,12 @@ def imgaug_transform(cfg: DictConfig) -> iaa.Sequential:
     else:
         raise NotImplementedError("must choose imgaug kind from 'default', 'dlc', 'dlc-top-down'")
 
-    # always apply resizing transform
-    data_transform.append(
-        iaa.Resize({
-            "height": cfg.data.image_resize_dims.height,
-            "width": cfg.data.image_resize_dims.width}
-        ))
+    # do not resize when using dynamic crop pipeline
```
[the remark this answers is not in the document]
Comment (Selmaan) [reply lightning_pose/data/augmentations.py:100]: This is a little tricky to handle, because we haven't totally fleshed out the way the pipeline is going to work? But I think we are planning to have just one dataset, in which case we will want access to the full, un-resized images, although I'm not sure yet how to do this in an efficient way keeping the images on the GPU

File: lightning_pose/data/dali.py [hunk 7]
```diff
@@ -155,8 +157,12 @@ def _dali_output_to_tensors(batch: list) -> UnlabeledBatchDict:
         frames = batch[0]["frames"][0, :, :, :, :]
         # shape (1,) or (2, 3)
         transforms = batch[0]["transforms"][0]
+        # get frame size, order is batch,seq_len,H,W,C
```
[the remark this answers is not in the document]
Comment (Selmaan) [reply lightning_pose/data/dali.py:160]: thx, yea I copied that from DALI documentation but I think you're right, will change

File: scripts/configs/config_birdCOM.yaml (hunk 5 above)
Comment (Selmaan) [reply scripts/configs/config_birdCOM.yaml:32]: that is weird and something I never compared! In my old code I had a L-R flip but not a top-down flip. Interested to see what matters

File: scripts/configs/config_birdCOM_backup.yaml
[the remark this answers is not in the document]
Comment (Selmaan) [reply scripts/configs/config_birdCOM_backup.yaml:1]: I'll delete this in a future commit

File: lightning_pose/models/heatmap_tracker.py [hunk 8]
```diff
@@ -269,16 +269,40 @@ def forward(
         # softmax temp stays 1 here; to modify for model predictions, see constructor
         return spatial_softmax2d(heatmaps, temperature=torch.tensor([1.0]))
 
+    def convert_bbox_coords(self, batch_dict: Union[HeatmapLabeledBatchDict, UnlabeledBatchDict],
```
[the remark this answers is not in the document]
Comment (Selmaan) [reply lightning_pose/models/heatmap_tracker.py:272]: will make this change

[the remark this answers is not in the document]
Comment (themattinthehatt) [reply lightning_pose/models/heatmap_tracker.py:272]: sounds good - can also pass those in as args

the main thing is i want a unit test for this function, since it's an important one

File: lightning_pose/models/heatmap_tracker.py (hunk 4 above)
Comment (Selmaan) [reply lightning_pose/models/heatmap_tracker.py:289]: unsqueeze works!

File: lightning_pose/models/heatmap_tracker.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Selmaan) [reply lightning_pose/models/heatmap_tracker.py]: will do that next. all other changes should be fixed in the new commit below

**Comment by themattinthehatt** [comment]
closes #86
